File: Persistencia/Scripts/micro_dao/servico.py
from micro_dao.objeto_modelo import ObjetoModelo

class Servico(ObjetoModelo):

    _id: int
    _valor: float
    _titulo: str
    _fk_usuario: int
    _fk_tipo_servico: int
    _fk_subtipo_servico: int
    _fk_contrato: int

    def __init__(self, valor: float, titulo: str, fk_usuario: int,
                 fk_tipo_servico: int, fk_subtipo_servico: int, fk_contrato: int):
        self.set_valor(valor)
        self.set_titulo(titulo)
        self.set_fk_usuario(fk_usuario)
        self.set_fk_tipo_servico(fk_tipo_servico)
        self.set_fk_subtipo_servico(fk_subtipo_servico)
        self.set_fk_contrato(fk_contrato)

    def get_id(self) -> int:
        return self._id

    def set_id(self, id: int):
        self._id = id
        return self

    def get_valor(self) -> float:
        return self._valor

    def set_valor(self, valor: float):
        self._valor = valor
        return self

    def get_titulo(self) -> str:
        return self._titulo

    def set_titulo(self, titulo: str):
        self._titulo = titulo
        return self

    def set_fk_usuario(self, fk_usuario: int):
        self._fk_usuario = fk_usuario
        return self

    def get_fk_usuario(self) -> int:
        return self._fk_usuario

    def set_fk_tipo_servico(self, fk_tipo_servico: int):
        self._fk_tipo_servico = fk_tipo_servico
        return self

    def get_fk_tipo_servico(self) -> int:
        return self._fk_tipo_servico

    def set_fk_subtipo_servico(self, fk_subtipo_servico: int):
        self._fk_subtipo_servico= fk_subtipo_servico
        return self

    def get_fk_subtipo_servico(self) -> int:
        return self._fk_subtipo_servico

    def set_fk_contrato(self, fk_contrato: int):
        self._fk_contrato = fk_contrato
        return self

    def get_fk_contrato(self) -> int:
        return self._fk_contrato

File: Projeto_Java/src/Dominio/Entidades/Servico.java
package Dominio.Entidades;

import Dominio.Enum.ESubtipoServico;
import Dominio.Enum.ETipoServico;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "servico")
public class Servico implements Comparable<Servico>{

    @Id
    @Column(name = "servico_id")
    private int id;

    @Column
    private String titulo;

    @Column
    private String descricao;

    @Column
    private ETipoServico tipo;

    private List<ESubtipoServico> subtipos;

    @Column
    private double valor;

    @Column(name = "fk_usuario")
    private int fkUsuario;

    @Column(name = "fk_tipo_servico")
    private int fkTipoServico;

    @Column(name = "fk_contrato")
    private int fkContrato;

    @ManyToMany
    @JoinTable(
            name="servico_subtipo_servico",
            joinColumns={@JoinColumn(name="servico_id")},
            inverseJoinColumns={@JoinColumn(name="subtipo_servico_id")}
            )
    private List<SubtipoServico> subtiposServico;

    public Servico() {}

    public Servico(int id, String titulo, String descricao, double valor, int fkTipoServico,
                   int fkUsuario, int fkContrato) {
        this.setId(id);
        this.setTitulo(titulo);
        this.setDescricao(descricao);
        this.setValor(valor);
        this.setFkTipoServico(fkTipoServico);
        this.setFkUsuario(fkUsuario);
        this.setFkContrato(fkContrato);
    }

    public Servico(String titulo, String descricao, List<ESubtipoServico> subtipos,
                   double valor, int fkTipoServico, int fkUsuario, int fkContrato) {
        this.setTitulo(titulo);
        this.setDescricao(descricao);
        this.setSubtipoServico(subtipos);
        this.setValor(valor);
        this.setFkTipoServico(fkTipoServico);
        this.setFkUsuario(fkUsuario);
        this.setFkContrato(fkContrato);
    }

    public Servico(int id, String titulo, String descricao,
                   List<ESubtipoServico> subtipos, double valor, int fkTipoServico,
                   int fkUsuario, int fkContrato) {
        this.setId(id);
        this.setTitulo(titulo);
        this.setDescricao(descricao);
        this.setSubtipoServico(subtipos);
        this.setValor(valor);
        this.setFkTipoServico(fkTipoServico);
        this.setFkUsuario(fkUsuario);
        this.setFkContrato(fkContrato);
    }


    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitulo() {
        return titulo;
    }

    public void setTitulo(String titulo) {

        if (!Util.isStringValida(titulo)) Util.throwExceptCampoVazio("título");

        this.titulo = titulo;
    }

    public String getDescricao() {
        return descricao;
    }

    public void setDescricao(String descricao) {

        if (!Util.isStringValida(descricao)) Util.throwExceptCampoVazio("descrição");

        this.descricao = descricao;
    }

    public double getValor() {
        return valor;
    }

    public void setValor(double valor) {

        if (valor < 0) Util.throwExceptNumeroInferior("preço", 0);

        this.valor = valor;
    }

    public ETipoServico getTipo() {
        return tipo;
    }

    public void setTipo(ETipoServico tipo) {
        this.tipo = tipo;
    }

    public List<ESubtipoServico> getSubtipos() {
        return subtipos;
    }

    public void setSubtipoServico(List<ESubtipoServico> subtipos) {
        this.subtipos = subtipos;
    }

    public void addSubtipoServico(ESubtipoServico subtipo) {

        if (this.subtipos == null) {
            this.subtipos = new ArrayList<>();
        }

        this.subtipos.add(subtipo);
    }

    public int getFkUsuario() {
        return fkUsuario;
    }

    private void setFkUsuario(int fkUsuario) {
        this.fkUsuario = fkUsuario;
    }

    public int getFkTipoServico() {
        return fkTipoServico;
    }

    public void setFkTipoServico(int fkTipoServico) {
        this.fkTipoServico = fkTipoServico;
        this.tipo = ETipoServico.getById(fkTipoServico);
    }

    public int getFkContrato() {
        return fkContrato;
    }

    public void setFkContrato(int fkContrato) {
        this.fkContrato = fkContrato;
    }

    @Override
    public String toString() {

        String servicoStr = "\nID:\t\t\t\t\t" + getId();
        servicoStr += "\nTítulo:\t\t\t\t" + getTitulo();
        servicoStr += "\nDescrição:\t\t\t" + getDescricao();
        servicoStr += "\nTipo:\t\t\t\t" + getTipo().getTipo();

        for (int i = 0; i < getSubtipos().size(); i++) {
            servicoStr += String.format("\nSubtipo %d:\t\t\t%s", i+1, getSubtipos().get(i).getSubtipo());
        }

        servicoStr += "\nValor:\t\t\t\t" + getValor();
        servicoStr += "\nFK_usuario:\t\t\t" + getFkUsuario();
        servicoStr += "\nFK_tipo_servico:\t" + getFkTipoServico();
        servicoStr += "\nFK_contrato:\t\t" + getFkContrato();

        return servicoStr;
    }


    @Override
    public int compareTo(Servico servico) {
        return Integer.compare(getId(), servico.getId());
    }
}
